src/bin/server.rs

```diff
@@ -183,38 +192,132 @@ impl Collector for CollectorServer {
         let mut resp = Vec::with_capacity(req.end - req.start);
         for (client_index, values_share, verifier_share, eval_proof) in results.into_iter() {
             resp.push((verifier_share, eval_proof));
-            coll.aggregate_by_attributes_state
+            coll.attribute_based_metrics_state
                 .insert(client_index, values_share);
         }
 
         resp
     }
 
-    async fn aggregate_by_attributes_finish(
+    async fn attribute_based_metrics_result(
         self,
         _: context::Context,
-        req: AggregateByAttributesResultRequest,
+        req: AttributeBasedMetricsResultRequest,
     ) -> Vec<Vec<Field128>> {
         debug_assert!(req.start < req.end);
         let mut coll = self.arc.lock().unwrap();
 
         for rejected_client_index in req.rejected {
             debug_assert!(coll
-                .aggregate_by_attributes_state
+                .attribute_based_metrics_state
                 .remove(&rejected_client_index)
                 .is_some());
         }
 
         let mut agg_share =
             vec![vec![Field128::zero(); coll.mastic.input_len()]; req.num_attributes];
-        for (_clinet_index, values_share) in coll.aggregate_by_attributes_state.iter() {
+        for (_clinet_index, values_share) in coll.attribute_based_metrics_state.iter() {
             for (a, v) in agg_share.iter_mut().zip(values_share.iter()) {
                 vec_add(a, v);
             }
         }
 
         agg_share
     }
+
+    async fn plain_metrics_validate(
+        self,
+        _: context::Context,
+        req: PlainMetricsValidateRequest,
+    ) -> Vec<Vec<u8>> {
+        debug_assert!(req.start < req.end);
+        let mut coll = self.arc.lock().unwrap();
+        let mut results = Vec::with_capacity(req.end - req.start);
+        let agg_id = self.server_id.try_into().unwrap();
+        let chunk_length = (coll.mastic.input_len() as f64).sqrt() as usize;
+        let prio3 = Prio3::new_histogram(2, coll.mastic.input_len(), chunk_length).unwrap();
+
+        results.par_extend((req.start..req.end).into_par_iter().map(|client_index| {
+            let ReportShare::Prio3 {
+                nonce,
+                public_share_bytes,
+                input_share_bytes,
+            } = &coll.report_shares[client_index].1
+            else {
```

Comment:
What kind of sorcery is this `else` without and `if`?

Reply:
hahahahha! It's a useful pattern under certain conditions. Basically rust let's you unpack things with a `let` statement, e.g.:

```rust
let SomeStruct{ field1, field2 } = some_struct;
```

You can do the same with enums, but if the variant you get is unexpected, then you need to handle it. You can use `else {}` for this:

```rust
let SomeEnum::ExpectedVariant{ field1, field2 } = some_enum else {
     // handle unexpected variant
};
```

Reply:
Thanks for explaining, yeah I've seen the `let` pattern but I didn't know you can use an else after it, cool!